### PR TITLE
MLP: Multilayer Perceptron Phase 2

### PR DESCRIPTION
**Pull request by cooper-sloan**
JIRA: MADLIB-1134

Weights, warm start, n_tries,
regularization, learning rate policy,
standardization and tests.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/141/

File: doc/design/modules/neural-network.tex [hunk 1]
```diff
@@ -22,15 +22,16 @@
...
 \end{moduleinfo}
 
 % Abstract. What is the problem we want to solve?
 This module implements artificial neural network \cite{ann_wiki}.
 
 \section{Multilayer Perceptron}
 Multilayer perceptron is arguably the most popular model among many neural network models \cite{mlp_wiki}.
-Here, we learn the coefficients by minimizing a least square objective function (\cite{bertsekas1999nonlinear}, example 1.5.3).
+Here, we learn the coefficients by minimizing a least square objective function, or cross entropy (\cite{bertsekas1999nonlinear}, example 1.5.3).
+The parallel architecture is based on the paper by Zhihen Huang \cite{mlp_parallel}.
```
Comment (haying) [inline doc/design/modules/neural-network.tex]: "Zhiheng" instead

Comment (cooper-sloan) [inline doc/design/modules/neural-network.tex]: Good catch.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/142/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/143/

**Comment by njayaram2** [comment]
LGTM

File: doc/design/modules/neural-network.tex [hunk 2]
```diff
@@ -46,41 +47,49 @@ \subsection{Formal Description}
 In the remaining part of this section, we will give a formal description of the derivation of objective function and its gradient.
 
 \paragraph{Objective function.}
-We mostly follow the notations in example 1.5.3 from Bertsekas \cite{bertsekas1999nonlinear}, for a multilayer perceptron that has $N$ layers (stages), and the $k$th stage has $n_k$ activation units ($\phi : \mathbb{R} \to \mathbb{R}$), the objective function is given as
-\[f_{(y, z)}(u) = \frac{1}{2} \|h(u, y) - z\|_2^2,\]
-where $y \in \mathbb{R}^{n_0}$ is the input vector, $z \in \mathbb{R}^{n_N}$ is the output vector,
+We mostly follow the notations in example 1.5.3 from Bertsekas \cite{bertsekas1999nonlinear}, for a multilayer perceptron that has $N$ layers (stages), and the $k$th stage has $n_k$ activation units ($\phi : \mathbb{R} \to \mathbb{R}$), the objective function for regression is given as
+\[f_{(x, y)}(u) = \frac{1}{2} \|h(u, x) - y\|_2^2,\]
```
Comment (njayaram2) [inline doc/design/modules/neural-network.tex]: `$k$th` -> `$k$^{th}`.

File: doc/design/modules/neural-network.tex [hunk 3]
```diff
@@ -46,41 +47,49 @@ \subsection{Formal Description}
...
+\[u_{k}^{sj} = uniform(-r,r),\]
+where r is defined as follows:
+\[r = \sqrt{\frac{6}{n_k+n_{k+1}}}\]
+With regularization, an additional term enters the objective function, given as
+\[\sum_{u_k^{sj}} \frac{1}{2} \lambda u_k^{sj2} \]
 This still leaves $h : \mathbb{R}^{n_0} \to \mathbb{R}^{n_N}$ as an open item.
-Let $x_k \in \mathbb{R}^{n_k}, k = 1,...,N$ be the output vector of the $k$th layer. Then we define $h(u, y) = x_N$, based on setting $x_0 = y$ and the $j$th component of $x_k$ is given in an iterative fashion as
-\footnote{$x_k^0 \equiv 1$ is used to simplified the notations, and $x_k^0$ is not a component of $x_k$, for any $k = 0,...,N$.}
+Let $o_k \in \mathbb{R}^{n_k}, k = 1,...,N$ be the output vector of the $k$th layer. Then we define $h(u, x) = o_N$, based on setting $o_0 = x$ and the $j$th component of $o_k$ is given in an iterative fashion as
+\footnote{$o_k^0 \equiv 1$ is used to simplified the notations, and $o_k^0$ is not a component of $o_k$, for any $k = 0,...,N$.}
```
Comment (njayaram2) [inline doc/design/modules/neural-network.tex]: `$j$th` -> `$j$^{th}`.
Similar changes in several other places.

File: doc/design/modules/neural-network.tex [hunk 4]
```diff
@@ -46,41 +47,49 @@ \subsection{Formal Description}
...
-\[f_{(y, z)}(u) = \frac{1}{2} \|h(u, y) - z\|_2^2,\]
-where $y \in \mathbb{R}^{n_0}$ is the input vector, $z \in \mathbb{R}^{n_N}$ is the output vector,
+We mostly follow the notations in example 1.5.3 from Bertsekas \cite{bertsekas1999nonlinear}, for a multilayer perceptron that has $N$ layers (stages), and the $k$th stage has $n_k$ activation units ($\phi : \mathbb{R} \to \mathbb{R}$), the objective function for regression is given as
+\[f_{(x, y)}(u) = \frac{1}{2} \|h(u, x) - y\|_2^2,\]
+and for classification the objective function is given as
+\[f_{(x, y)}(u) = \sum_i (\log(h_i(u, x)) * z_i + (1-\log(h_i(u, x))) *( 1- z_i) ,\]
+where $x \in \mathbb{R}^{n_0}$ is the input vector, $y \in \mathbb{R}^{n_N}$ is the output vector (one hot encoded for classification),
 \footnote{Of course, the objective function can be defined over a set of input-output vector pairs, which is simply given as the addition of the above $f$.}
 and the coefficients are given as
```
Comment (njayaram2) [inline doc/design/modules/neural-network.tex]: Change ```classification),\n
\footnote{```
to ```classification),~\footnote{```

Similar comment for the second footnote too.

File: doc/design/modules/neural-network.tex [hunk 5]
```diff
@@ -46,41 +47,49 @@ \subsection{Formal Description}
...
-    &= \left( h^t(u, y) - z^t \right) \cdot \frac{\partial h^t(u, y)}{\partial u_{N-1}^{st}} \\
-    &= \left( x_N^t - z^t \right) \cdot \frac{\partial x_N^t}{\partial u_{N-1}^{st}} \\
-    &= \left( x_N^t - z^t \right) \cdot \frac{\partial \phi \left( \sum_{s=0}^{n_{N-1}} x_{N-1}^s u_{N-1}^{st} \right)}{\partial u_{N-1}^{st}} \\
-    &= \left( x_N^t - z^t \right) \cdot \phi' \left( \sum_{s=0}^{n_{N-1}} x_{N-1}^s u_{N-1}^{st} \right) \cdot x_{N-1}^s \\
+    &= \left( h^t(u, x) - y^t \right) \cdot \frac{\partial h^t(u, x)}{\partial u_{N-1}^{st}} \\
+    &= \left( o_N^t - y^t \right) \cdot \frac{\partial o_N^t}{\partial u_{N-1}^{st}} \\
+    &= \left( o_N^t - y^t \right) \cdot \frac{\partial \phi \left( \sum_{s=0}^{n_{N-1}} o_{N-1}^s u_{N-1}^{st} \right)}{\partial u_{N-1}^{st}} \\
+    &= \left( o_N^t - y^t \right) \cdot \phi' \left( \sum_{s=0}^{n_{N-1}} o_{N-1}^s u_{N-1}^{st} \right) \cdot o_{N-1}^s \\
 \end{aligned}\]
 To ease the notation, let the input vector of the $j$th activation unit of the $(k+1)$th layer be
```
Comment (njayaram2) [inline doc/design/modules/neural-network.tex]: `$(k+1)$th` -> `$(k+1)$^{th}`.

File: doc/design/modules/neural-network.tex [hunk 6]
```diff
@@ -197,3 +207,44 @@ \subsubsection{The $\mathit{Gradient}$ Function}
...
+        \State $Y_j \set \texttt{subset-of-Y}$
+    \EndFor
+    \For{$i = 1,...,t$}
+        \For{$j = 1,...,s$}
+            \State $u_j \set copy(u)$
+            \State $u_j \set \texttt{mlp-train-iteration}(X_j, Y_j, \eta)$
+        \EndFor
+        \State $u \set \texttt{weighted-avg}(u_{1...s})$
+    \EndFor
+    \State \Return $u$
```
Comment (njayaram2) [inline doc/design/modules/neural-network.tex]: Just the `return u` shows up in a new page in the pdf, see if you can avoid that.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/145/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/146/

File: src/ports/postgres/modules/convex/test/mlp.sql_in [hunk 7]
```diff
@@ -241,9 +247,8 @@ SELECT mlp_predict(
 
 select * from mlp_prediction;
 SELECT assert(
-    -- Accuracy greater than 90%
-    COUNT(*)/150.0 > 0.95,
-    'MLP: Accuracy is too low (< 95%). Wrong result.'
+    COUNT(*)/150.0 > 0.99,
```
Comment (orhankislal) [inline src/ports/postgres/modules/convex/test/mlp.sql_in]: Tested on PG 9.4 and the latest commit of GPDB 5 (1 node, 3 segments).
IC passes on Postgres but fails on GPDB5. I tested the code manually and it works fine with 148/150 accuracy. I would suggest dropping the assertion threshold to 95% to make sure such minute changes do not fail the whole test.

File: src/ports/postgres/modules/convex/test/mlp.sql_in [hunk 8]
```diff
@@ -241,9 +247,8 @@ SELECT mlp_predict(
 
 select * from mlp_prediction;
```
Comment (orhankislal) [inline src/ports/postgres/modules/convex/test/mlp.sql_in]: select * from -> SELECT * FROM

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/150/

**Comment by orhankislal** [comment]
GPDB 5 IC passes as well. LGTM

**Pull request closed without merging**